### PR TITLE
deviceplugin: generalize A && B variants

### PR DESCRIPTION
**Pull request by fromanirh**
Extract both variants in a more generic plugin, which is now
configurable with a JSON config file.
The configuration file allows to specify which devices should be created on which hostname, with special value '*' meaning "every host".
Also adds example config file

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
Docs are most likely lacking. Please suggest what needs to be documented better.

File: config.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+				"resource": "example.com/deviceB",
+				"id": "DevB1",
+				"healthy": true,
+				"numanode": 1
+			},
+			{
+				"resource": "example.com/deviceB",
+				"id": "DevB2",
+				"healthy": true,
+				"numanode": 1
```
Comment (swatisehgal) [inline config.json:32]: NIT: To match the diagram in Readme, this will have to be 0 as well

**Comment by swatisehgal** [comment]
A few suggestions below:

- Updating the README to point to the example config and indicate that it would create a setup such as the diagram included. 

- Update the Dockerfile and/or manifests to include the command line

- Minor formatting changes made to the environment variable in the last commit, can we include that?

- Minor update to the example config file is needed to match the diagram in the README

File: config.json (hunk 1 above)
Comment (fromanirh) [reply config.json:32]: ack, done

**Comment by fromanirh** [comment]
force pushed because of a rebase :\
PENDING:
- fixing the environment variable name as requested
- proper testing against a real cluster

**Comment by fromanirh** [comment]
> force pushed because of a rebase :
> PENDING:
> 
>     * fixing the environment variable name as requested
> 
>     * proper testing against a real cluster

env var naming fixed

File: config/device_B.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			},
+			{
+				"id": "DevB2",
+				"healthy": true,
+				"numanode": 0
+			},
+			{
+				"id": "DevB3",
+				"healthy": true,
+				"numanode": 1`
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply config/device_B.json:17]: fixed

File: config/devices.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			},
+			{
+				"id": "DevB2",
+				"healthy": true,
+				"numanode": 0
+			},
+			{
+				"id": "DevB3",
+				"healthy": true,
+				"numanode": 1`
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply config/devices.json:32]: fixed

**Comment by fromanirh** [comment]
all fixed and tested, AFAICT seems good. I had to revert some changes, so I decided to squash all the patches.

**Comment by swatisehgal** [comment]
I tested this PR, I noticed that in a scenario when we request both device type A and type B it creates a single device file /dev/tty10. Previous implementation would create tty10, tty11... corresponding to request for device type A and tty20, tty21 corresponding to request for device type B. A solution of this could be we make device files configurable parameter too but open to alternatives

**Comment by swatisehgal** [comment]
Looks good, tested it works as expected!

**Comment by swatisehgal** [comment]
Looks good, works as expected!

- I have submitted a PR to your branch suggesting minor changes [here](https://github.com/fromanirh/sample-device-plugin/pull/1). It is based on your branch so it should merge into your code without issues

- Rebase you branch on the latest master branch and squash commits

Thanks for this excellent PR!

**Comment by fromanirh** [comment]
PR merged and changes rebased + squashed

**Pull request merged**